CRM/Sqltasks/Task.php

```diff
@@ -337,9 +338,15 @@ public function execute($params = []) {
       $this->status = 'error';
       $this->log("Task is still running. Execution skipped.");
       return $this->log_messages;
+    } elseif ($this->isTaskLock()) {
```

Comment:
This essentially calls `$lock->isAcquired()` before `$lock->acquire()` is called, so we don't hold the lock yet when this code runs. This causes all task executions to fail with `Task is locked. Execution skipped.`.

The correct usage/order for locking looks like this:

```php
$lock = new CRM_Core_Lock('lock_name');
$lock->acquire();
if (!$lock->isAcquired()) {
  // task is locked
}
else {
  // run
  $lock->release();
}
```

Reply:
To fix that I use `$lock->isFree()`.The method gets fresh value from database.
It works much faster then `acquire() + isAcquired()` methods.
( ~0.3s vs ~2s - checking time when current task is locked)
And it works in any orders of methods calls.

Now it looks like:
```php
if ($task->isTaskLock()) {
  // task is locked 
} else {
  $task->lockTask();
  //run task
  $task->unlockTask();
}
```
The same using raw core locker:
```php
$lock = new CRM_Core_Lock('lock_name');
if (!$lock->isFree()) {
  // task is locked
}
else {
  $lock->acquire();
  // run task
  $lock->release();
}
```
But `$lock->isFree()` method is marked as deprecated in `LockInterface`. But method implementation is written and works well. 
This case has a lot of pluses, that's why I use `$lock->isFree()` method.
It's okay?

Reply:
@awg-github note that this code also suffers from a race condition. Two threads could be checking `$lock->isFree()` at the same time, and since the lock would be free up until `$lock->acquire()` is called later on, the check would pass for both threads.

`$lock->acquire()` would then create a lock for whichever thread reaches that code first, and the second thread would not get a lock, but would continue after attempting to obtain the lock for two seconds (the code does not check the return value of `$lock->acquire()` or use `$lock->isAcquired()` to see if the lock was actually obtained).

I think the order mentioned in my previous comment would work best here - it's less complex/easier to think about, and the performance impact here is negligible: the slow code path with a timeout of two seconds would only be triggered if the *first* check based on `running_since` runs into a race condition - which is very rare.